packages/plugin-core/src/commands/AddExistingVaultCommand.ts

```diff
@@ -0,0 +1,504 @@
+          if (select?.title === openDocsOption) {
+            // Open a page in the default browser that describes what transitive
+            // dependencies are, and how to add them.
+            await PluginFileUtils.openWithDefaultApp(
+              "https://wiki.dendron.so/notes/q9yo0y7czv8mxlkbnw1ugj1"
+            );
+          }
+        }
+      }
+    } catch (err) {
+      // If anything does fail, ignore the error. This check is not crucial to
+      // adding a vault, it's better if we let the user keep adding.
+      Logger.warn({
+        ctx: "VaultAddCommand.handleRemoteRepoSelfContained",
+        err,
+      });
+    }
+  }
+
+  async addWorkspaceToWorkspace(workspace: DWorkspace) {
+    const wsRoot = this._ext.getDWorkspace().wsRoot;
+    const vaults = workspace.vaults;
+    // Some things, like updating workspace file, can't be parallelized so needs to be done one at a time
+    for (const vault of vaults) {
+      // eslint-disable-next-line no-await-in-loop
+      await this.addVaultToWorkspace(vault);
+    }
+    // add to gitignore
+    await GitUtils.addToGitignore({
+      addPath: workspace.name,
+      root: wsRoot,
+      noCreateIfMissing: true,
+    });
+
+    const workspaceDir = path.join(wsRoot, workspace.name);
+    await fs.ensureDir(workspaceDir);
+    await GitUtils.addToGitignore({
+      addPath: ".dendron.cache.*",
+      root: workspaceDir,
+    });
+  }
+
+  async addVaultToWorkspace(vault: DVault) {
+    return WorkspaceUtils.addVaultToWorkspace({
+      vault,
+      wsRoot: this._ext.getDWorkspace().wsRoot,
+    });
+  }
+
+  /**
+   * Returns all vaults added
+   * @param opts
+   * @returns
+   */
+  async execute(opts: CommandOpts) {
+    const ctx = "AddExistingVaultCommand";
+    let vaults: DVault[] = [];
+    Logger.info({ ctx, msg: "enter", opts });
+    if (opts.type === VaultType.REMOTE) {
+      if (opts.isSelfContained) {
+        ({ vaults } = await this.handleRemoteRepoSelfContained(opts));
+      } else {
+        ({ vaults } = await this.handleRemoteRepo(opts));
+      }
+    } else {
+      const wsRoot = this._ext.getDWorkspace().wsRoot;
+      const fsPath = VaultUtils.normVaultPath({
+        vault: { fsPath: opts.path },
+        wsRoot,
+      });
+      const wsService = new WorkspaceService({ wsRoot });
+      const vault: DVault = {
+        fsPath,
+      };
+      // Make sure these don't get set to undefined, or serialization breaks
+      if (await fs.pathExists(path.join(opts.path, FOLDERS.NOTES))) {
+        vault.selfContained = true;
+      }
+      if (opts.name) {
+        vault.name = opts.name;
+      }
+
+      if (VaultUtils.isSelfContained(vault)) {
+        await wsService.createSelfContainedVault({
+          vault,
+          addToConfig: true,
+          addToCodeWorkspace: false,
+          newVault: false,
+        });
+      } else {
+        await wsService.createVault({ vault });
+      }
+      await this.addVaultToWorkspace(vault);
```

Comment:
this part has a lot of overlap with `CreateNewVault`. can we combine some of this common logic in a `VaultUtils` class?

Reply:
This didn't work out as helpful while combining because `wsService` and other common part that is present has dependency on `engine-server` which is not a dependency in common-all(VaultUtils)